File: src/main/java/io/tripled/social/client/application/DefaultPostMessageUseCase.java
package io.tripled.social.client.application;

import io.tripled.social.client.domain.DateTimeProvider;
import io.tripled.social.client.domain.Message;
import io.tripled.social.client.domain.SocialNetwork;
import io.tripled.social.client.domain.SocialNetworkRepository;
import io.tripled.social.client.domain.UserName;

public class DefaultPostMessageUseCase implements PostMessageUseCase {

  private SocialNetworkRepository socialNetworkRepository;
  private DateTimeProvider dateTimeProvider;

  public DefaultPostMessageUseCase(SocialNetworkRepository socialNetworkRepository, DateTimeProvider dateTimeProvider) {
    this.socialNetworkRepository = socialNetworkRepository;
    this.dateTimeProvider = dateTimeProvider;
  }

  @Override
  public String postMessage(String userName, String message) {
    SocialNetwork socialNetwork = socialNetworkRepository.get();

    Message msg = new Message(new UserName(userName), message, dateTimeProvider);

    return socialNetwork.postMessage(msg);
  }
}

File: src/main/kotlin/io/tripled/social/client/application/DefaultPostMessageUseCase.kt
package io.tripled.social.client.application

import io.tripled.social.client.domain.DateTimeProvider
import io.tripled.social.client.domain.SocialNetwork
import io.tripled.social.client.domain.SocialNetworkRepository
import io.tripled.social.client.domain.UserName

class DefaultPostMessageUseCase(socialNetworkRepository: SocialNetworkRepository, dateTimeProvider: DateTimeProvider) :
    PostMessageUseCase {
    private val socialNetworkRepository: SocialNetworkRepository
    private val dateTimeProvider: DateTimeProvider

    init {
        this.socialNetworkRepository = socialNetworkRepository
        this.dateTimeProvider = dateTimeProvider
    }

    override fun postMessage(userName: kotlin.String, message: kotlin.String): kotlin.String {
        val socialNetwork: SocialNetwork = socialNetworkRepository.get()

        val msg = io.tripled.social.client.domain.Message(UserName(userName), message, dateTimeProvider)

        return socialNetwork.postMessage(msg)
    }
}
